File: apps/dimi/src/tui/components/messages/user-message.ts
/**
 * Renders a user message in the transcript.
 */

import { Spacer, Text, truncateToWidth, visibleWidth, type Component } from '@dimi-agent/pi-tui';

import { ImageThumbnail } from '#/tui/components/media/image-thumbnail';
import { USER_MESSAGE_BULLET } from '#/tui/constant/symbols';
import { currentTheme } from '#/tui/theme';
import type { ImageAttachment } from '#/tui/utils/image-attachment-store';
import { isRenderCacheEnabled } from '#/tui/utils/render-cache';

export class UserMessageComponent implements Component {
  private text: string;
  private readonly bullet?: string;
  private spacerComponent: Spacer;
  private imageThumbnails: ImageThumbnail[];

  private renderCache: { width: number; lines: string[] } | undefined;

  constructor(text: string, images?: ImageAttachment[], bullet?: string) {
    this.text = text;
    this.bullet = bullet;
    this.spacerComponent = new Spacer(1);
    this.imageThumbnails = images?.map((img) => new ImageThumbnail(img)) ?? [];
  }

  private markRenderDirty(): void {
    this.renderCache = undefined;
  }

  invalidate(): void {
    this.markRenderDirty();
    for (const img of this.imageThumbnails) {
      img.invalidate?.();
    }
  }

  render(width: number): string[] {
    const safeWidth = Math.max(0, width);
    if (safeWidth <= 0) return [''];

    if (
      isRenderCacheEnabled() &&
      this.renderCache !== undefined &&
      this.renderCache.width === safeWidth
    ) {
      return this.renderCache.lines;
    }

    const marker = this.bullet ?? USER_MESSAGE_BULLET;
    const bullet = marker.length > 0 ? currentTheme.boldFg('roleUser', marker) : '';
    const bulletWidth = visibleWidth(bullet);
    const contentWidth = Math.max(1, safeWidth - bulletWidth);

    const lines: string[] = [];

    // Spacer
    for (const line of this.spacerComponent.render(safeWidth)) {
      lines.push(line);
    }

    // Text is re-dyed from the current theme; invalidate() (theme change) clears
    // the render cache so the new colours are picked up on the next render.
    const coloredText = currentTheme.boldFg('roleUser', this.text);
    const textLines = new Text(coloredText, 0, 0).render(contentWidth);
    for (let i = 0; i < textLines.length; i++) {
      const prefix = i === 0 ? bullet : ' '.repeat(bulletWidth);
      lines.push(prefix + textLines[i]);
    }

    // Images — indented to align with text after the bullet
    for (const thumbnail of this.imageThumbnails) {
      const imageLines = thumbnail.render(contentWidth);
      for (const line of imageLines) {
        lines.push(' '.repeat(bulletWidth) + line);
      }
    }

    const rendered = lines.map((line) => {
      // Inline image sequences (Kitty / iTerm2) carry their own placement
      // information and have zero visible width, but pi-tui's truncateToWidth
      // treats the embedded base64 payload as visible text and would chop the
      // escape sequence in half, leaving garbage like "0m...". Skip truncation
      // for those lines; the image itself already respects maxWidthCells.
      if (isImageLine(line)) return line;
      return truncateToWidth(line, safeWidth, '…');
    });
    if (isRenderCacheEnabled()) {
      this.renderCache = { width: safeWidth, lines: rendered };
    }
    return rendered;
  }
}

function isImageLine(line: string): boolean {
  return line.includes('\u001B_G') || line.includes('\u001B]1337;File=');
}

/**
 * Invisible turn-boundary marker for replay. Some replayed records start a
 * new turn without anything to show — the goal driver's synthetic
 * continuation prompt is model-facing and never rendered live — but the
 * transcript still needs a mounted boundary component so step/assistant
 * folding (and window trimming) can find the turn edges. Renders zero lines.
 */
export class ReplayTurnBoundaryComponent implements Component {
  invalidate(): void {}
  render(_width: number): string[] {
    return [];
  }
}

File: crates/dimi-tui/src/components/messages/user_message.rs
//! `UserMessageComponent` — renders a user message in the transcript
//! (port of `apps/dimi/src/tui/components/messages/user-message.ts`).

use crate::component::Component;
use crate::components::messages::USER_MESSAGE_BULLET;
use crate::components::text::Text;
use crate::theme::{ColorToken, current_theme};
use crate::width::visible_width;
use crate::wrap::truncate_to_width;

/// Renders a user message: a role-colored bullet + the message text, wrapped
/// with continuation indent after the bullet.
pub struct UserMessageComponent {
    text: String,
    bullet: Option<String>,
    render_cache: Option<(usize, Vec<String>)>,
}

impl UserMessageComponent {
    pub fn new(text: &str, bullet: Option<String>) -> Self {
        UserMessageComponent {
            text: text.to_owned(),
            bullet,
            render_cache: None,
        }
    }

    pub fn set_text(&mut self, text: &str) {
        self.text = text.to_owned();
        self.render_cache = None;
    }
}

impl Component for UserMessageComponent {
    fn render(&mut self, width: usize) -> Vec<String> {
        let safe_width = width;
        if safe_width == 0 {
            return vec![String::new()];
        }
        if let Some((cached_width, cached_lines)) = &self.render_cache {
            if *cached_width == safe_width {
                return cached_lines.clone();
            }
        }

        let marker = self
            .bullet
            .clone()
            .unwrap_or_else(|| USER_MESSAGE_BULLET.to_owned());
        let bullet = if marker.is_empty() {
            String::new()
        } else {
            current_theme().bold_fg(ColorToken::RoleUser, &marker)
        };
        let bullet_width = visible_width(&marker);
        let content_width = (safe_width.saturating_sub(bullet_width)).max(1);

        let mut lines: Vec<String> = Vec::new();
        // Spacer(1).
        lines.push(String::new());

        let colored_text = current_theme().bold_fg(ColorToken::RoleUser, &self.text);
        let mut text_component = Text::new(&colored_text, 0, 0);
        let text_lines = text_component.render(content_width);
        for (i, text_line) in text_lines.iter().enumerate() {
            let prefix = if i == 0 {
                bullet.clone()
            } else {
                " ".repeat(bullet_width)
            };
            lines.push(format!("{prefix}{text_line}"));
        }

        let rendered: Vec<String> = lines
            .iter()
            .map(|line| truncate_to_width(line, safe_width, "…", false))
            .collect();
        self.render_cache = Some((safe_width, rendered.clone()));
        rendered
    }

    fn invalidate(&mut self) {
        self.render_cache = None;
    }
}

/// Invisible turn-boundary marker (replay) — renders zero lines.
pub struct ReplayTurnBoundaryComponent;

impl Component for ReplayTurnBoundaryComponent {
    fn render(&mut self, _width: usize) -> Vec<String> {
        Vec::new()
    }
    fn invalidate(&mut self) {}
}
